devcodeabode/MoDiBo
# Add plugin manager

**Outstanding Dependencies**

- [x] #8 for configuration integration
- [x] Finalizing blueprint plugin structure

**Related bugs/feature requests**



+ Resolves  #5 

**Describe the changes**

Adds a plugin manager to load in plugins from the file system.

**Expected behavior**

Enabled plugins will be loaded and sorted into their types.

**New npm dependencies**


- `glob:^8.0.3`

**Testing Instructions**

 1. Create a new folder in `plugins/`
 2. In that folder, create `main.js` with these contents:
  ```js
const utils = require("../../src/utils.js");
const configManager = require("../../src/configManager.js");
const NAME = "Test Plugin";
const SLUG = "test-plugin";
const AUTHOR = "Some Person";
const SHORT_HELP = `${NAME} by ${AUTHOR}: Help text here!`;
const COMMANDS = ["command1", "command2"];
function help(command = null, args = null) {
  return "How to use the command with given args.";
}
function processCommand(command, args, bot, message) {
  utils.logger.log("debug", `From ${NAME}: Command ${command} ; Args ${args}`);
}
function processMessage(bot, message) {
  utils.logger.log("debug", `From ${NAME}: Message ${message.content}`);
}
function startCron(bot) {
  utils.logger.log("debug", `From ${NAME}: Started Cron`);
}
function onLoad(bot) {
  utils.logger.log("debug", `${NAME} has started.`);
}

module.exports = {
  NAME,
  SLUG,
  SHORT_HELP,
  AUTHOR,
  COMMANDS,
  help,
  processCommand,
  processMessage,
  startCron,
  onLoad,
};

```
3. Repeat with a second and third test plugin. Change the `NAME` and `SLUG`, and comment out the exports to one or more of the functions so each has a different combination. Also change the `COMMANDS` for each copy of the plugin.
4. Make sure the console output level for your testing bot is at `debug`
5. In your `config.json`, add in empty objects for each test plugin. For example...
```json
  "plugins": {
    "test-plugin": {},
    "test-plugin-2": {},
    "test-plugin-3": {}
  },
```
6. Run the testing bot. Check the console output for successful and correct loading.
7. Send a message with one of the commands. Verify you see...
  - All Message plugins log a debug saying they got the message
  - The Command plugin with the corresponding command logs a debug with the command details
9. Remove one of the test plugins from the config and re-run the bot. Verify that only the plugins in the config load.
10. Remove the `plugins` key from the config completely. Verify that a warning is generated an no plugins load.

**Additional context**
